<a href="https://colab.research.google.com/github/SimphiweMahaye/2022-South-African-Developer-Survey/blob/main/1_Data_Collection%2C_Normalising_%26_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Data Collection, Normalising & Wrangling**



The dataset used is a csv file collected from the url: https://insights.stackoverflow.com/survey

The description of the file is:

With over 70,000 responses fielded from over 180 countries, our Annual Developer Survey examines all aspects of the developer experience from learning to code to their favorite technologies to version control and the workplace experience of professional developers.

I first start by importing my libraries, Pandas which assist us into saving the csv file into a dataframe and Google import since I'm importing it from the local disk into my Google Colaboratory Notebook.

In [ ]:
import pandas
import numpy
from google.colab import files

In [ ]:
uploaded = files.upload()


Saving survey_results_public.csv to survey_results_public.csv


I then convert my uploaded csv into a Pandas dataframe.

In [ ]:
complete_dataframe = pandas.read_csv('survey_results_public.csv')
complete_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73268 entries, 0 to 73267
Data columns (total 79 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ResponseId                      73268 non-null  int64  
 1   MainBranch                      73268 non-null  object 
 2   Employment                      71709 non-null  object 
 3   RemoteWork                      58958 non-null  object 
 4   CodingActivities                58899 non-null  object 
 5   EdLevel                         71571 non-null  object 
 6   LearnCode                       71580 non-null  object 
 7   LearnCodeOnline                 50685 non-null  object 
 8   LearnCodeCoursesCert            29389 non-null  object 
 9   YearsCode                       71331 non-null  object 
 10  YearsCodePro                    51833 non-null  object 
 11  DevType                         61302 non-null  object 
 12  OrgSize                         

After running the info method and from the output, I can see that it contains the 'Country' column. We had known that from the data schema, and since we are only interested in the country South Africa, we will then filter our dataframe to only show us a condensed dataframe only showing respondents from South Africa. We also note the total number of entries(Rows) which is 73268, so if our conversion table worked, it's safe to assume that the number of those entries will decrease.

In [ ]:
south_africa_dataframe = complete_dataframe.loc[complete_dataframe['Country'] == 'South Africa']
south_africa_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 493 entries, 100 to 72706
Data columns (total 79 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ResponseId                      493 non-null    int64  
 1   MainBranch                      493 non-null    object 
 2   Employment                      493 non-null    object 
 3   RemoteWork                      430 non-null    object 
 4   CodingActivities                430 non-null    object 
 5   EdLevel                         493 non-null    object 
 6   LearnCode                       493 non-null    object 
 7   LearnCodeOnline                 336 non-null    object 
 8   LearnCodeCoursesCert            232 non-null    object 
 9   YearsCode                       489 non-null    object 
 10  YearsCodePro                    378 non-null    object 
 11  DevType                         443 non-null    object 
 12  OrgSize                         

I now have a total of 493 entries, which is safe to say my table has been filtered down. I will then show the coultry column to show if all the columns are of the country South Africa.

In [ ]:
south_africa_dataframe['Country'].unique()

array(['South Africa'], dtype=object)

There are no other countries present. I now proceed to removing any missing values, removing any duplicated then re index my list to have some cohesive structure.

In [ ]:
south_africa_dataframe.drop_duplicates(subset=['ResponseId'])
south_africa_dataframe = south_africa_dataframe.dropna(subset=['ResponseId'])
south_africa_dataframe = south_africa_dataframe2.reset_index(drop=True)
south_africa_dataframe.head()

ResponseId                      MainBranch  \
0         101     I code primarily as a hobby   
1         343  I am a developer by profession   
2         625  I am a developer by profession   
3        1035  I am a developer by profession   
4        1074  I am a developer by profession   

                                          Employment    RemoteWork  \
0                                 Student, full-time           NaN   
1                                Employed, full-time  Fully remote   
2  Employed, full-time;Independent contractor, fr...  Fully remote   
3  Employed, full-time;Independent contractor, fr...  Fully remote   
4                                Employed, full-time  Fully remote   

                           CodingActivities  \
0                                       NaN   
1                                     Hobby   
2  Hobby;Contribute to open-source projects   
3            Hobby;Bootstrapping a business   
4  Hobby;Contribute to open-source projects   

                                           EdLevel  \
0                        Primary/elementary school   
1     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   

                                           LearnCode  \
0  Other online resources (e.g., videos, blogs, f...   
1  Other online resources (e.g., videos, blogs, f...   
2  Books / Physical media;Other online resources ...   
3  Books / Physical media;Other online resources ...   
4  Books / Physical media;Friend or family member...   

                                     LearnCodeOnline LearnCodeCoursesCert  \
0  Blogs;Written Tutorials;Stack Overflow;Video-b...                  NaN   
1  Technical documentation;Programming Games;Writ...                  NaN   
2  Technical documentation;Blogs;Written Tutorial...                  NaN   
3  Technical documentation;Blogs;Written Tutorial...                  NaN   
4  Technical documentation;Written Tutorials;Stac...                  NaN   

  YearsCode  ...               TimeSearching        TimeAnswering  \
0        11  ...                         NaN                  NaN   
1        30  ...         30-60 minutes a day  30-60 minutes a day   
2        20  ...                         NaN                  NaN   
3         7  ...                         NaN                  NaN   
4        26  ...  Less than 15 minutes a day  30-60 minutes a day   

      Onboarding                                   ProfessionalTech  \
0            NaN                                                NaN   
1      Very long                      DevOps function;Microservices   
2            NaN                                                NaN   
3            NaN                                                NaN   
4  Somewhat long  DevOps function;Microservices;Developer portal...   

  TrueFalse_1 TrueFalse_2 TrueFalse_3           SurveyLength  \
0         NaN         NaN         NaN  Appropriate in length   
1          No          No         Yes              Too short   
2         NaN         NaN         NaN               Too long   
3         NaN         NaN         NaN  Appropriate in length   
4         Yes          No         Yes  Appropriate in length   

                   SurveyEase ConvertedCompYearly  
0                        Easy                 NaN  
1                        Easy             56897.0  
2  Neither easy nor difficult            115000.0  
3                        Easy             39449.0  
4                        Easy            164370.0  

[5 rows x 79 columns]

I download this simplified dataset for further analysis and visualizations.

In [216]:
south_africa_dataframe.to_csv('SouthAfricaDataSet.csv', encoding = 'utf-8-sig') 
files.download('SouthAfricaDataSet.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Since the first question we want to answer is; **"What programming language should I learn first?"**  Our primary concern are the columns:

1.   LanguageHaveWorkedWith
2.   LanguageWantToWorkWith


1.   DatabaseHaveWorkedWith
2.   DatabaseWantToWorkWith


1.   PlatformHaveWorkedWith
2.   PlatformWantToWorkWith

The issue with those columns is the data entries are all in a single column, I would need to split the data entries to be in individual colums for our visualisations and plots to work.

In [ ]:
LanguageHaveWorkedWith_split_languages = south_africa_dataframe.LanguageHaveWorkedWith.str.split(';', expand = True)
LanguageHaveWorkedWith_split_languages.head()

0         1           2           3           4     5     6     7   \
0         APL       C++      Python        None        None  None  None  None   
1  Bash/Shell  HTML/CSS  JavaScript         PHP        Rust   SQL  None  None   
2         C++  HTML/CSS  JavaScript        Rust  TypeScript  None  None  None   
3        Dart  HTML/CSS        Java  JavaScript         SQL  None  None  None   
4  Bash/Shell   Clojure    HTML/CSS        Java  JavaScript   SQL  None  None   

     8     9     10    11    12    13    14    15  
0  None  None  None  None  None  None  None  None  
1  None  None  None  None  None  None  None  None  
2  None  None  None  None  None  None  None  None  
3  None  None  None  None  None  None  None  None  
4  None  None  None  None  None  None  None  None

From the table above, I've split our values and assigned them arbitrarily for the purpose of counting the number of times each word appears on the database and that would give us our values to plot with.

In [ ]:
flat_list = LanguageHaveWorkedWith_split_languages.values.flatten().tolist()
flat_list
LanguageHaveWorkedWithCount = pandas.Series(flat_list).value_counts()
LanguageHaveWorkedWithCount.columns=["Programming Language Users Have Worked With", "Number of Users"]
LanguageHaveWorkedWithCount.dropna()
LanguageHaveWorkedWithCount.head()

JavaScript    333
SQL           287
HTML/CSS      286
Python        200
C#            177
dtype: int64

I have now asigned a simplified dataframe, with one column being the Programming language and the next column being the number of users that have worked with said language.

The next step now is to save the dataframe in a csv format and then do the same with the rest of our required columns from the survey.

In [ ]:
LanguageHaveWorkedWithCount.to_csv('LanguageHaveWorkedWith.csv', encoding = 'utf-8-sig') 
files.download('LanguageHaveWorkedWith.csv')


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
lwtww_split = south_africa_dataframe.LanguageWantToWorkWith.str.split(';', expand = True)
lwtww_flat = lwtww_split.values.flatten().tolist()
lwtww_count = pandas.Series(lwtww_flat).value_counts()

lwtww_count.to_csv('LanguageWantToWorkWith.csv', encoding = 'utf-8-sig') 
files.download('LanguageWantToWorkWith.csv')


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
dbhww_split = south_africa_dataframe.DatabaseHaveWorkedWith.str.split(';', expand = True)
dbhww_flat = dbhww_split.values.flatten().tolist()
dbhww_count = pandas.Series(dbhww_flat).value_counts()

dbhww_count.to_csv('DatabaseHaveWorkedWith.csv', encoding = 'utf-8-sig') 
files.download('DatabaseHaveWorkedWith.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
dbwtww_split = south_africa_dataframe.DatabaseWantToWorkWith.str.split(';', expand = True)
dbwtww_flat = dbwtww_split.values.flatten().tolist()
dbwtww_count = pandas.Series(dbwtww_flat).value_counts()

dbwtww_count.to_csv('DatabaseWantToWorkWith.csv', encoding = 'utf-8-sig') 
files.download('DatabaseWantToWorkWith.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
phww_split = south_africa_dataframe.PlatformHaveWorkedWith.str.split(';', expand = True)
phww_flat = phww_split.values.flatten().tolist()
phww_count = pandas.Series(phww_flat).value_counts()

phww_count.to_csv('PlatformHaveWorkedWith.csv', encoding = 'utf-8-sig') 
files.download('PlatformHaveWorkedWith.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
pwtww_split = south_africa_dataframe.PlatformWantToWorkWith.str.split(';', expand = True)
pwtww_flat = pwtww_split.values.flatten().tolist()
pwtww_count = pandas.Series(pwtww_flat).value_counts()

pwtww_count.to_csv('PlatformWantToWorkWith.csv', encoding = 'utf-8-sig') 
files.download('PlatformWantToWorkWith.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Now that I have collected my data, it's time to move onto the fun part, analysing it and getting the answers to the questions I had been asking.